cli/shell/kubectl.go

```diff
@@ -0,0 +1,131 @@
+	versionString := string(bytes)
+	if err != nil {
+		return [3]int{}, fmt.Errorf("Error running kubectl Version. Output: %s Error: %v", versionString, err)
+	}
+
+	if err != nil {
+		return version, err
+	}
+
+	justTheVersionString := strings.TrimPrefix(versionString, "Client Version: v")
+	justTheMajorMinorRevisionNumbers := strings.Split(justTheVersionString, "-")[0]
+	split := strings.Split(justTheMajorMinorRevisionNumbers, ".")
+
+	if len(split) < 3 {
+		return version, fmt.Errorf("Unknown Version string format from Kubectl: [%s] not enough segments: %s", versionString, split)
+	}
+
+	for i, segment := range split {
+		v, err := strconv.Atoi(strings.TrimSpace(segment))
+		if err != nil {
+			return version, fmt.Errorf("Unknown Version string format from Kubectl: [%s], not an integer: [%s]", versionString, segment)
+		}
+		version[i] = v
+	}
+
+	return version, nil
+}
+
+func (kctl *kubectl) StartProxy(potentialErrorWhenStartingProxy chan error, port int) error {
+	fmt.Printf("Running `kubectl proxy %d`\n", port)
+
+	if kctl.ProxyPort() != portWhenProxyNotRunning {
+		return fmt.Errorf("Kubectl proxy already running on port [%d]", kctl.ProxyPort)
+	}
+
+	output, err := kctl.sh.AsyncStdout(potentialErrorWhenStartingProxy, "kubectl", "proxy", "-p", strconv.Itoa(port))
+
+	kubectlOutput, err := kctl.sh.WaitForCharacter(magicCharacterThatIndicatesProxyIsRunning, output, kubectlDefaultTimeout)
+
+	fmt.Println(kubectlOutput)
+	if err != nil {
+		return fmt.Errorf("Error waiting for kubectl to start the proxy. kubetl returned [%s], error: %v", kubectlOutput, err)
+	}
+
+	kctl.proxyPort = kubectlDefaultProxyPort
+	return nil
+}
+
+func (kctl *kubectl) UrlFor(namespace string, extraPathStartingWithSlash string) (string, error) {
+	if kctl.ProxyPort() == portWhenProxyNotRunning {
+		return "", errors.New("proxy needs to be started before generating URLs")
+	}
+
+	url := fmt.Sprintf("http://%s:%d/api/v1/namespaces/%s%s", "127.0.0.1", kctl.ProxyPort(), namespace, extraPathStartingWithSlash)
+	return url, nil
+}
+
+func isCompatibleVersion(minimalRequirementVersion [3]int, actualVersion [3]int) bool {
+	if minimalRequirementVersion[0] < actualVersion[0] {
+		return true
+	}
+
+	if (minimalRequirementVersion[0] == actualVersion[0]) && minimalRequirementVersion[1] <= actualVersion[1] {
+		return true
+	}
+
+	if (minimalRequirementVersion[0] == actualVersion[0]) && (minimalRequirementVersion[1] == actualVersion[1]) && (minimalRequirementVersion[2] <= actualVersion[2]) {
+		return true
+	}
+
+	return false
+}
+
+func MakeKubectl(shell Shell) (Kubectl, error) {
+
+	kubectl := &kubectl{
+		sh:        shell,
+		proxyPort: portWhenProxyNotRunning,
+	}
+
+	actualVersion, err := kubectl.Version()
+
+	if err != nil {
+		return nil, err
+	}
+
+	if !isCompatibleVersion(minimunKubectlVersionExpected, actualVersion) {
+		return nil, fmt.Errorf("Kubectl is on version %v, but version %v or more recent is required", actualVersion, minimunKubectlVersionExpected)
```

Comment:
TIOLI -- you could clean this up a bit if you added a helper to convert int[3] to version string. otherwise, as written, you'll end up with:

"Kubectl is on version [1 2 3], but version [2 3 4] or more recent is required"